File: sample-projects/swift/Sources/CellarsAndCentaurs/Domain/Model/Dice.swift
struct Dice {
    let sides: Int

    func roll() -> DiceRoll {
        DiceRoll(dice: self, value: Int.random(in: 1...sides))
    }
}

struct DiceRoll {
    let dice: Dice
    let value: Int

    var isCriticalHit: Bool { value == dice.sides }
}

func rollD20() -> DiceRoll {
    let d20Roll = Dice(sides: 20).roll()
    return d20Roll
}

File: sample-projects/kotlin/src/main/kotlin/de/sots/cellarsandcentaurs/domain/model/Dice.kt
package de.sots.cellarsandcentaurs.domain.model

import kotlin.random.Random

const val DEFAULT_DIE_SIDES = 20

data class Dice(val sides: Int = DEFAULT_DIE_SIDES) {
    fun roll(): DiceRoll = DiceRoll(Random.nextInt(1, sides + 1))
}

data class DiceRoll(val d20Roll: Int) {
    val isCritical: Boolean
        get() = d20Roll == DEFAULT_DIE_SIDES
}

fun rollD20(): Int = Dice().roll().d20Roll
